File: frontend/src/lib/m365-mail.ts
/** Microsoft 365 — correo operativo (Outlook) */

export type M365MailFolder = "inbox" | "sent" | "drafts" | "archive";

export interface M365AccountSummary {
  id: string;
  email: string | null;
  display_name: string | null;
  connection_mode: string;
  connection_status: string;
  microsoft_user_id?: string | null;
  token_expires_at?: string | null;
  last_sync_at?: string | null;
  is_active: boolean;
  jaios_user_id?: string | null;
  token_status?: string | null;
  last_graph_error?: string | null;
}

export interface M365ConnectionState {
  connected: boolean;
  account_connected: boolean;
  azure_configured: boolean;
  oauth_ready: boolean;
  read_only: boolean;
  message: string;
  tenant_id: string | null;
  client_id: string | null;
  redirect_uri: string;
  webhook_url: string | null;
  connected_accounts: M365AccountSummary[];
  active_account: M365AccountSummary | null;
}

export interface M365MailAttachment {
  id?: string | null;
  name: string;
  size_bytes?: number | null;
  content_type?: string | null;
  is_inline?: boolean;
}

export interface M365MailMessage {
  id: string;
  subject: string;
  sender: string;
  sender_name?: string | null;
  to_recipients: string[];
  cc_recipients?: string[];
  received_at: string | null;
  sent_at?: string | null;
  preview: string;
  body_html?: string | null;
  body_text?: string | null;
  is_read: boolean;
  has_attachments: boolean;
  attachments?: M365MailAttachment[];
  web_link?: string | null;
  folder: string;
  account_id?: string | null;
  account_email?: string | null;
}

export interface M365MailListResponse {
  items: M365MailMessage[];
  total: number;
  folder: string;
  account_id?: string | null;
  account_email?: string | null;
  connected: boolean;
  read_only: boolean;
  message: string;
  error?: string;
  permission_hint?: string;
}

export interface M365MailDetail extends M365MailMessage {
  connected: boolean;
  read_only: boolean;
  can_reply: boolean;
  can_forward: boolean;
}

export const M365_MAIL_FOLDERS: { id: M365MailFolder; label: string }[] = [
  { id: "inbox", label: "Bandeja de entrada" },
  { id: "sent", label: "Enviados" },
  { id: "drafts", label: "Borradores" },
  { id: "archive", label: "Archivo" },
];

export const M365_GRAPH_PERMISSIONS = [
  { scope: "User.Read", description: "Perfil del usuario", adminConsent: false },
  { scope: "offline_access", description: "Refresh token", adminConsent: false },
  { scope: "Mail.Read", description: "Leer correos", adminConsent: false },
  { scope: "Mail.ReadWrite", description: "Marcar leído, archivar", adminConsent: false },
  { scope: "Mail.Send", description: "Enviar, responder, reenviar", adminConsent: false },
  { scope: "Calendars.Read", description: "Leer calendario", adminConsent: false },
  { scope: "Calendars.ReadWrite", description: "Crear/editar eventos", adminConsent: false },
  { scope: "Contacts.Read", description: "Leer contactos", adminConsent: false },
  { scope: "Contacts.ReadWrite", description: "Crear/editar contactos", adminConsent: false },
  { scope: "Files.Read", description: "Leer OneDrive", adminConsent: false },
  { scope: "Files.ReadWrite", description: "Subir archivos OneDrive", adminConsent: true },
  { scope: "Sites.Read.All", description: "Leer SharePoint", adminConsent: true },
  { scope: "Sites.ReadWrite.All", description: "Escribir SharePoint", adminConsent: true },
  { scope: "Team.ReadBasic.All", description: "Listar Teams", adminConsent: false },
  { scope: "Channel.ReadBasic.All", description: "Canales Teams", adminConsent: false },
  { scope: "ChannelMessage.Read.All", description: "Leer mensajes Teams", adminConsent: true },
  { scope: "Chat.Read", description: "Leer chats", adminConsent: true },
  { scope: "Chat.ReadWrite", description: "Enviar chats", adminConsent: true },
  { scope: "Group.Read.All", description: "Grupos Microsoft 365", adminConsent: true },
];

File: backend/app/schemas/m365_mail.py
"""Schemas — correo Microsoft 365 (Outlook operativo)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class M365MailAttachment(BaseModel):
    id: str | None = None
    name: str = ""
    size_bytes: int | None = None
    content_type: str | None = None
    is_inline: bool = False


class M365MailAttachmentRef(BaseModel):
    """Adjunto para envío — archivo local (base64) o referencia OneDrive."""

    name: str
    content_type: str = "application/octet-stream"
    content_base64: str | None = None
    source_url: str | None = None
    onedrive_item_id: str | None = None


class M365MailMessage(BaseModel):
    id: str
    subject: str = ""
    sender: str = ""
    sender_name: str | None = None
    to_recipients: list[str] = Field(default_factory=list)
    cc_recipients: list[str] = Field(default_factory=list)
    received_at: datetime | None = None
    sent_at: datetime | None = None
    preview: str = ""
    body_html: str | None = None
    body_text: str | None = None
    is_read: bool = False
    has_attachments: bool = False
    attachments: list[M365MailAttachment] = Field(default_factory=list)
    web_link: str | None = None
    folder: str = "inbox"
    account_id: uuid.UUID | None = None
    account_email: str | None = None


class M365MailListResponse(BaseModel):
    items: list[M365MailMessage]
    total: int
    folder: str
    account_id: uuid.UUID | None = None
    account_email: str | None = None
    connected: bool = False
    read_only: bool = False
    message: str = ""
    error: str | None = None
    permission_hint: str | None = None


class M365MailDetailResponse(M365MailMessage):
    connected: bool = True
    read_only: bool = False
    can_reply: bool = True
    can_forward: bool = True
    can_send: bool = True


class M365MailComposeRequest(BaseModel):
    body: str
    subject: str | None = None
    to: list[str] = Field(default_factory=list)
    cc: list[str] = Field(default_factory=list)
    comment: str | None = None
    attachments: list[M365MailAttachmentRef] = Field(default_factory=list)
    reply_all: bool = False
    save_draft: bool = False


class M365MailActionResponse(BaseModel):
    ok: bool
    message: str
    message_id: str | None = None


class M365MailPatchRequest(BaseModel):
    is_read: bool | None = None
    move_to_folder: str | None = None


class M365MailAccountsResponse(BaseModel):
    items: list
    total: int
    azure_configured: bool = False
    oauth_ready: bool = False
